src/raft.rs

```diff
@@ -1425,7 +1428,15 @@ impl<T: Storage> Raft<T> {
                         new_message(m.from, vote_resp_msg_type(m.get_msg_type()), None);
                     to_send.reject = true;
                     to_send.term = self.term;
+                    let (commit, commit_term) = self.raft_log.commit_info();
+                    to_send.commit = commit;
+                    to_send.commit_term = commit_term;
                     self.r.send(to_send, &mut self.msgs);
+                    // It should be safe to commit log by vote in all conditions. Here
+                    // use a stricter condition for best safety.
+                    if self.leader_id == INVALID_ID || !self.promotable {
+                        self.maybe_commit_by_vote(&m);
```

Comment:
If we assume quorum won't change during election, the candidate may get a `RequestVote` and forward the commit index. I suggest the check of new conf change can be written under `maybe_commit_by_vote`.
Also, I don't think the `if self.leader_id == INVALID_ID || !self.promotable` can improve the safety. Instead, it adds some dependency that is not necessary, such that the `leader_id` must be set to `INVALID_ID` after some time. How about just using `self.maybe_commit_by_vote` or, more simply, calling this function whatever this peer rejects the vote request or not because it has no side effect in all conditions.

Reply:
The condition can prevent follower in lease from updating its commit index by non-leader's message. What I try to do here is only updating commit index when there is no valid leader.

Reply:
You don't need to worry about this case. The code here will ignore the RequestVote msg when it still in lease.
https://github.com/tikv/raft-rs/blob/a45c4a33229090a0984638bfa4bc163ef8b284ea/src/raft.rs#L1256-L1287